File: internal_testsuites/golang/testsuite/startosis_remote_package_test/startosis_remote_package_test.go
package startosis_remote_package_test

import (
	"context"
	"github.com/kurtosis-tech/kurtosis-cli/golang_internal_testsuite/test_helpers"
	"github.com/sirupsen/logrus"
	"github.com/stretchr/testify/require"
	"testing"
)

const (
	testName              = "package"
	isPartitioningEnabled = false
	defaultDryRun         = false
	remotePackage         = "github.com/kurtosis-tech/datastore-army-package"
	executeParams         = `{"num_datastores": 2}`
	dataStoreService0Name = "datastore-0"
	dataStoreService1Name = "datastore-1"
	datastorePortId       = "grpc"
	defaultParallelism    = 4
)

func TestStartosisRemotePackage(t *testing.T) {
	ctx := context.Background()

	// ------------------------------------- ENGINE SETUP ----------------------------------------------
	enclaveCtx, destroyEnclaveFunc, _, err := test_helpers.CreateEnclave(t, ctx, testName, isPartitioningEnabled)
	require.NoError(t, err, "An error occurred creating an enclave")
	defer destroyEnclaveFunc()

	// ------------------------------------- TEST RUN ----------------------------------------------
	logrus.Debugf("Executing Starlark Package: '%v'", remotePackage)

	runResult, err := enclaveCtx.RunStarlarkRemotePackageBlocking(ctx, remotePackage, executeParams, defaultDryRun, defaultParallelism)
	require.NoError(t, err, "Unexpected error executing starlark package")

	require.Nil(t, runResult.InterpretationError, "Unexpected interpretation error. This test requires you to be online for the read_file command to run")
	require.Empty(t, runResult.ValidationErrors, "Unexpected validation error")
	require.Empty(t, runResult.ExecutionError, "Unexpected execution error")
	logrus.Infof("Successfully ran Starlark Package")

	// Check that the service added by the script is functional
	logrus.Infof("Checking that services are all healthy")
	require.NoError(
		t,
		test_helpers.ValidateDatastoreServiceHealthy(context.Background(), enclaveCtx, dataStoreService0Name, datastorePortId),
		"Error validating datastore server '%s' is healthy",
		dataStoreService0Name,
	)
	require.NoError(
		t,
		test_helpers.ValidateDatastoreServiceHealthy(context.Background(), enclaveCtx, dataStoreService1Name, datastorePortId),
		"Error validating datastore server '%s' is healthy",
		dataStoreService1Name,
	)
	logrus.Infof("All services added via the package work as expected")
}

File: internal_testsuites/typescript/src/testsuite/startosis_remote_package_test/startosis_remote_package.test.ts
import log from "loglevel"
import { err, ok, Result } from "neverthrow";

import { createEnclave } from "../../test_helpers/enclave_setup";
import {
    validateDataStoreServiceIsHealthy,
} from "../../test_helpers/test_helpers";

const TEST_NAME = "module"

const REMOTE_PACKAGE = "github.com/kurtosis-tech/datastore-army-package"
const EXECUTE_PARAMS            = `{"num_datastores": 2}`
const DATASTORE_SERVICE_0_ID     = "datastore-0"
const DATASTORE_SERVICE_1_ID   = "datastore-1"
const DATASTORE_PORT_ID       = "grpc"
const IS_DRY_RUN = false

const IS_PARTITIONING_ENABLED  = false

jest.setTimeout(180000)

test.skip("Test remote Starlark package execution", async () => {
    // ------------------------------------- ENGINE SETUP ----------------------------------------------
    const createEnclaveResult = await createEnclave(TEST_NAME, IS_PARTITIONING_ENABLED)

    if(createEnclaveResult.isErr()) { throw createEnclaveResult.error }

    const { enclaveContext, stopEnclaveFunction } = createEnclaveResult.value


    try {
        // ------------------------------------- TEST SETUP ----------------------------------------------
        log.info(`Executing Starlark package: '${REMOTE_PACKAGE}'`)
        const runResult = await enclaveContext.runStarlarkRemotePackageBlocking(REMOTE_PACKAGE, EXECUTE_PARAMS, IS_DRY_RUN)
        if (runResult.isErr()) {
            log.error("An error occurred executing the Starlark Package")
            throw runResult.error
        }

        expect(runResult.value.interpretationError).toBeUndefined()
        expect(runResult.value.validationErrors).toEqual([])
        expect(runResult.value.executionError).toBeUndefined()
        log.info("Successfully ran Starlark Package")

        log.info("Checking that services are all healthy")
        const validationResultDataStore0 = await validateDataStoreServiceIsHealthy(enclaveContext, DATASTORE_SERVICE_0_ID, DATASTORE_PORT_ID);
        if (validationResultDataStore0.isErr()) {
            throw err(new Error(`Error validating that service '${DATASTORE_SERVICE_0_ID}' is healthy`))
        }


        const validationResultDataStore1 = await validateDataStoreServiceIsHealthy(enclaveContext, DATASTORE_SERVICE_1_ID, DATASTORE_PORT_ID);
        if (validationResultDataStore1.isErr()) {
            throw err(new Error(`Error validating that service '${DATASTORE_SERVICE_1_ID}' is healthy`))
        }
        log.info("All services added via the package work as expected")
    }finally{
        stopEnclaveFunction()
    }

    jest.clearAllTimers()
})
